sys/include/net/gnrc/tcp.h

```diff
@@ -80,39 +110,33 @@ void gnrc_tcp_tcb_init(gnrc_tcp_tcb_t *tcb);
  *            -ETIMEDOUT if the connection could not be opened.
  *            -ECONNREFUSED if the connection was reset by the peer.
  */
-int gnrc_tcp_open_active(gnrc_tcp_tcb_t *tcb,  uint8_t address_family,
-                         char *target_addr, uint16_t target_port,
+int gnrc_tcp_open_active(gnrc_tcp_tcb_t *tcb, const gnrc_tcp_ep_t *remote,
                          uint16_t local_port);
```

Comment:
Why is `local_port` not becoming an endpoint?

Reply:
I would like to avoid using an Endpoint here just to specify the source port.

Using an endpoint here forces the user to store an endpoint on the stack, wasting memory (Unused IPv6 Address). In addition it is not selfexplaintory how to setup the Endpoint, using the parameter to specify the local port number is much clearer (at least from my point of view).

Reply:
I'm aware this is a behavioral change, but what if you want to bind your TCP connection to a specific local address (which is possible with `sock`)?

Reply:
I am not sure that this is a specified in the TCP RFC for an active connection opening (It is defined to passive open).

Reply:
Usually, if an RFC does not specify anything it is up to interpretation of the implementer. For Linux (using the python wrapper for simplicity sake) this works:

```py
import socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(("160.45.xxx.xxx", 1234)) # blacking out my public IP address here ;-)
s.connect(('127.0.0.1', 1337))
s.send(b"Hello world\n")
s.close()
```

With a `netcat` server running in parallel I get

```
$ nc -l 0.0.0.0 1337
Hello world
```

![image](https://user-images.githubusercontent.com/675644/72071037-2eacc780-32eb-11ea-9ed2-79724d0196a8.png)

Reply:
(this comment is non-blocking btw, as I know this involves a behavioral change).

Reply:
But if we want to make this possible with `gnrc_tcp`, it would be great if this door would already be opened with the according API change.

Reply:
Then the real question is: Do we need to implement that? Forcing a connection to use a specific source address on connection establishment seems to me like a rare use case. I prefer to let the IP-Layer decide that.

Reply:
Is this something the other implementations of sock are capable of?
If so it would make sence to implement just to be feature complete.

Reply:
> Is this something the other implementations of sock are capable of?

Triggered by that question I only just noticed, that `sock_tcp` does as exactly the same what you are proposing. Should have read the API before rambling instead of quoting from memory :sweat_smile: Sorry for the noise...

Reply:
This happens to everyone from time to time.